lib/luaxdp.c

```diff
@@ -48,43 +48,31 @@ __diag_ignore_all("-Wmissing-prototypes",
                   "Global kfuncs as their definitions will be in BTF");
 #endif
 
-static inline int luaxdp_callback(lua_State *L)
-{
-	if (lunatik_getregistry(L, luaxdp_callback) == LUA_TNIL) {
-		lua_pop(L, 1); /* pop nil */
-		if (lua_type(L, -1) != LUA_TFUNCTION) {
-			pr_err("couldn't find callback");
-			return -1;
-		}
-		lunatik_setregistry(L, -1, luaxdp_callback);
-	}
-	return 0;
-}
-
-static int luaxdp_handler(lua_State *L, lunatik_object_t *packet)
+static int luaxdp_handler(lua_State *L, lunatik_object_t *buffer, size_t offset)
 {
 	int action = -1;
 	int status;
 
-	if (luaxdp_callback(L) != 0)
+	if (lunatik_getregistry(L, luaxdp_handler) != LUA_TFUNCTION) {
+		pr_err("couldn't find callback");
 		goto out;
+	}
 
-	lunatik_pushobject(L, packet);
-	if ((status = lua_pcall(L, 1, 1, 0)) != LUA_OK) {
+	lunatik_pushobject(L, buffer);
+	lua_pushinteger(L, (lua_Integer)offset);
+	if ((status = lua_pcall(L, 2, 1, 0)) != LUA_OK) {
 		pr_err("%s\n", lua_tostring(L, -1));
-		goto pop;
+		goto out;
 	}
 
 	action = lua_tointeger(L, -1);
-pop:
-	lua_pop(L, 1);
 out:
 	return action;
 }
 
-__bpf_kfunc int bpf_luaxdp_run(struct xdp_md *xdp_ctx, char *key, size_t key__sz)
+__bpf_kfunc int bpf_luaxdp_run(struct xdp_md *xdp_ctx, char *key, size_t key__sz, size_t offset)
```

Comment:
does `offset` refer to the offset of packet data inside the buffer ?

Reply:
Yes, a offset within the packet, passed from the eBPF program.. I'm using currently for passing the payload offset, but it could be used by eBPF to point to any specific position of the packet.. my plan is to evolve this parm to pass a `void *` with a packed data from the eBPF program.. but, baby steps ;-)

Reply:
I can share a sample eBPF code to make it clearer if you think it's worth

Reply:
I understood the concept.